src/main/java/org/assertj/core/api/junit/jupiter/SoftAssertionsExtension.java

```diff
@@ -93,10 +114,131 @@
+    // find SoftAssertions fields in the test class hierarchy
+    Collection<Field> softAssertionsFields = findFields(testInstance.getClass(),
+                                                        field -> isAnnotated(field, InjectSoftAssertions.class),
+                                                        HierarchyTraversalMode.BOTTOM_UP);
+
+//    final List<SoftAssertionsProvider> providers = new ArrayList<>(softAssertionsFields.size());
+    for (Field field : softAssertionsFields) {
+
+      final int fieldModifiers = field.getModifiers();
+      if (Modifier.isStatic(fieldModifiers) || Modifier.isFinal(fieldModifiers)) {
+        throw new ExtensionConfigurationException(format("[%s] SoftAssertionsProvider field must not be static or final.",
+                                                         field.getName()));
+      }
+
+      Class<?> providerClass = (Class<?>) field.getType();
+      if (!SoftAssertionsProvider.class.isAssignableFrom(providerClass)) {
+        throw new ExtensionConfigurationException(format("[%s] field is not a SoftAssertionsProvider (%s).",
+                                                         field.getName(), providerClass.getTypeName()));
+      }
+      @SuppressWarnings("unchecked") // Guaranteed because of the above sanity check
+      final Class<? extends SoftAssertionsProvider> softAssertionsProvider = (Class<? extends SoftAssertionsProvider>) providerClass;
+
+      if (Modifier.isAbstract(providerClass.getModifiers())) {
+        throw new ExtensionConfigurationException(format("[%s] SoftAssertionsProvider [%s] is abstract and cannot be instantiated.",
+                                                         field.getName(), providerClass));
+      }
+      try {
+        providerClass.getDeclaredConstructor();
+      } catch (Exception e) {
+        throw new ExtensionConfigurationException(format("[%s] SoftAssertionsProvider [%s] does not have a default constructor",
+                                                         field.getName(), providerClass.getName()));
+      }
+
+      field.setAccessible(true);
+      SoftAssertionsProvider softAssertions = getSoftAssertionsProvider(context, softAssertionsProvider);
+      try {
+        field.set(testInstance, softAssertions);
+      } catch (IllegalAccessException e) {
+        throw new ExtensionConfigurationException(format("[%s] Could not gain access to field", field.getName()), e);
+      }
+    }
+  }
+
+  @Override
+  public void beforeEach(ExtensionContext context) throws Exception {
+    AssertionErrorCollector collector = getAssertionErrorCollector(context);
+
+    if (isPerClassConcurrent(context)) {
+      ThreadLocalErrorCollector tlec = getThreadLocalCollector(context);
+      tlec.setDelegate(collector);
+    } else {
+      while (initialiseDelegate(context, collector)) {
+        context = context.getParent().get();
+      }
```

Comment:
To fully understand this part, you need to read it in conjunction with `postProcessTestInstance()` and `instantiateProvider()`.

If a test is running `SAME_THREAD`, then concurrency is not a concern because you'll never have more than one thread accessing any of the fields or the soft assertion state. Likewise, if you're running `PER_METHOD`, then Jupiter creates a fresh test instance for each test, which again means that the individual tests will not interfere with each other's state. So the only time inter-test interference becomes an issue is if you're running PER_CLASS *and* CONCURRENT. We handle these two cases differently.

In the first case, where concurrency is not an issue, we perform the following steps:

1. `postProcessTestInstance()` calls `instantiateProvider()` for all of the valid `@InjectSoftAssertions`-annotated fields.
2. `instantiateProvider()` will instantiate the soft assertions provider of the required type. If the test method has already started, it will also set its delegate to the current `AssertionErrorCollector`.
3. `beforeEach()` will find all instantiated soft assertions providers and set their delegates to the current `AssertionErrorCollector`. It will also loop through the parent contexts (eg, if there is a nested class) and set the delegates for any fields they may have registered (so that nested fields will work).

So by the time the test starts running, all of the fields will be initialised and their delegates set to the current `AssertionErrorCollector`:

SoftAssertionsProvider field (for the current instance) -> AssertionErrorCollector (for the current test)

(where "->" reads as "delegates to")

In the second case, where concurrency is a concern, the following alternative code path happens:
1. `postProcessTestInstance()` calls `instantiateProvider()` for all of the valid `@InjectSoftAssertions`-annotated fields (this part is the same as above).
2. `instantiateProvider()` instantiates the soft assertions provider of the required type. This time, instead of setting the delegate to the one for the current context (which is at the class level in the PER_CLASS case before any individual tests have started), it sets it to a class-level `ThreadLocalErrorCollector`.
3. `beforeEach()` calls `setDelegate()` on the `ThreadLocalErrorCollector` with the AssertionErrorCollector for the current test context.

So by the time the test starts running, all of the fields will be initialised as:

SoftAssertionsProvider field (for the current instance) -> ThreadLocalErrorCollector (for the current instance) -> AssertionErrorCollector (for the current Thread, ie the current test)

So each thread will end up writing its assertions into a collector that is unique to the test it is current running, and not into some object with class-level scope.

Hopefully that explains it a bit better; I added some comments in the code to try and help others too.